### PR TITLE
Fix #224: Update text for static DH oracle

### LINKED ISSUE #224: Update static DH oracle text
From @hugokraw:

> An attack with Q queries, lowers security by (log Q)/2 bits (not log Q), and it only applies if Q is a divisor of p-1 or p+1. Moreover, the *memory* cost of the attack (not just time) is 2^{p/2-(log(Q)/2}. 

We might also want to update the "use a larger curve" guidance, since whether or not it's needed depends on the feasibility of these attacks. (OPAQUE, for example, is fine to use P-256 and the like.)

More information can be found in [Taylor's writeup](https://mumble.net/~campbell/tmp/20210116/draft-riastradh-staticdh.html) on  "Attacks on Static Diffie-Hellman Oracles".

### PR DESCRIPTION
**Pull request by alxdavids**
Apologies for the long wait on this, CI seems to be failing but it doesn't immediately look like a linting error?

### CODE REVIEW

File: draft-irtf-cfrg-voprf.md [hunk 1]
```diff
@@ -1333,19 +1333,24 @@ The assumption that this problem is hard was first introduced in
...
+divides p-1 or p+1, where p is the order of the prime-order group used
+to instantiate the OPRF. 
 
 As an example, suppose that a group instantiation is used that provides
 128 bits of security against discrete log cryptanalysis. Then an
 adversary with access to a Q-sDH oracle and makes Q=2^20 queries can
-reduce the security of the instantiation by log_2(2^20) = 20 bits.
+reduce the security of the instantiation by log_2(2^20)/2 = 10 bits. It
+is worth highlighting that launching an attack would require
+2^{p/2-(log(Q)/2} bits of memory.
```
Comment (chris-wood) [inline draft-irtf-cfrg-voprf.md:1345]:
```suggestion
reduce the security of the instantiation by log_2(2^20)/2 = 10 bits. Launching an attack would require
2^{p/2-(log(Q)/2} bits of memory.
```

File: draft-irtf-cfrg-voprf.md [hunk 2]
```diff
@@ -1356,16 +1361,19 @@ the ideal functionality provided by the PRF.
 The OPRF instantiations that we recommend in this document are informed
 by the cryptanalytic discussion above. In particular, choosing elliptic
 curves configurations that describe 128-bit group instantiations would
-appear to in fact instantiate an OPRF with 128-log_2(Q) bits of
-security.
+appear to in fact instantiate an OPRF with 128-(log_2(Q)/2) bits of
+security. Moreover, such attacks are only possible for those certain
+applications where the adversary can query the OPRF directly.
+In applcations where such an oracle is not made available, such as in
```
Comment (chris-wood) [inline draft-irtf-cfrg-voprf.md:1367]: Hmm... is this true? The OPRF oracle output is included in the server's response to the client, it just so happens that the OPRF key is unique per user.

File: draft-irtf-cfrg-voprf.md [hunk 3]
```diff
@@ -1356,16 +1361,19 @@ the ideal functionality provided by the PRF.
...
 launch a highly expensive attack to reduce security to anything much
 below 100 bits of security. We see this possibility as something that
-may result in problems in the future. For applications that cannot
-tolerate discrete logarithm security of lower than 128 bits, we
-recommend only implementing ciphersuites with IDs: 0x0002, 0x0004, and
-0x0005.
+may result in problems in the future. For applications that admit the
+aforementioned oracle functionality, and that cannot tolerate discrete logarithm
+security of lower than 128 bits, we recommend only implementing
+ciphersuites with IDs: 0x0002, 0x0004, and 0x0005.
```
Comment (chris-wood) [inline draft-irtf-cfrg-voprf.md:1376]:
```suggestion
security of lower than 128 bits, we recommend only implementing
ciphersuites with IDs 0x0002, 0x0004, and 0x0005.
```

**Review by chris-wood — COMMENTED** [review]
This LGTM pending one question about OPAQUE.

File: draft-irtf-cfrg-voprf.md (hunk 1 above)
Comment (armfazh) [reply draft-irtf-cfrg-voprf.md:1345]: it this 2^p of 2^log(p) ?

File: draft-irtf-cfrg-voprf.md (hunk 2 above)
Comment (alxdavids) [reply draft-irtf-cfrg-voprf.md]: Ah, misunderstood the setting, I have changed the text

File: draft-irtf-cfrg-voprf.md [hunk 4]
```diff
@@ -1356,16 +1360,20 @@ the ideal functionality provided by the PRF.
...
 by the cryptanalytic discussion above. In particular, choosing elliptic
 curves configurations that describe 128-bit group instantiations would
-appear to in fact instantiate an OPRF with 128-log_2(Q) bits of
-security.
+appear to in fact instantiate an OPRF with 128-(log_2(Q)/2) bits of
+security. Moreover, such attacks are only possible for those certain
+applications where the adversary can query the OPRF directly.
+In applications where such an oracle is not made available, or where
+OPRF keys are derived on a per-client basis such as in
+{{!I-D.irtf-cfrg-opaque}}, the security loss does not apply.
```
Comment (chris-wood) [inline draft-irtf-cfrg-voprf.md:1368]:
```suggestion
In applications where such an oracle is not made available this security loss does not apply.
```

**Review by chris-wood — APPROVED** [review]
Recommend removing the OPAQUE reference entirely. We can clarify the situation over in that document. Pending that change, this LGTM.

File: draft-irtf-cfrg-voprf.md [hunk 5]
```diff
@@ -1333,19 +1333,23 @@ The assumption that this problem is hard was first introduced in
...
+log_2(Q)/2 bits. Note that the attacks only work in situations where Q
+divides p-1 or p+1, where p is the order of the prime-order group used
+to instantiate the OPRF. 
 
 As an example, suppose that a group instantiation is used that provides
 128 bits of security against discrete log cryptanalysis. Then an
 adversary with access to a Q-sDH oracle and makes Q=2^20 queries can
-reduce the security of the instantiation by log_2(2^20) = 20 bits.
+reduce the security of the instantiation by log_2(2^20)/2 = 10 bits. Launching an attack would require
+2^{p/2-(log(Q)/2} bits of memory.
```
Comment (armfazh) [inline draft-irtf-cfrg-voprf.md:1344]:
```suggestion
2^(p/2-log_2(Q)/2) bits of memory.
```
Please double-check.

**Review by armfazh — APPROVED** [review]

**Comment by alxdavids** [comment]
LGTM; merging!

**Pull request merged**